File: src/mtglab/decks/suggest.py
"""Ranked replacement candidates for a card that has to leave a deck.

Two decks currently fail the gate on a banned card each, and ADR 8 says the
gate blocks rather than substituting, because picking the replacement is a
deckbuilding judgement that belongs to the deck's owner. This module does not
change that. It answers a narrower, mechanical question:

    of every card that is legal here, which ones most resemble the one being
    removed?

That is a measurement, not a recommendation, and the distinction is the whole
design. Nothing here decides anything; it produces a shortlist with the reason
each candidate scored, so a human can disagree with it quickly. Every fact it
uses comes from the corpus -- CLAUDE.md rule 1 -- so it cannot assert a card
does something it does not.

Scoring is deliberately pure Python over `CardRecord`s. Only `candidate_pool`
touches the database, which keeps the interesting half testable without one.

The score
---------
Similarity is a weighted sum, each part in 0..1:

    card type      0.30   replacing a creature with an instant changes the deck
    mana value     0.20   a 6-drop is not replaced by a 2-drop
    keywords       0.15   Scryfall's own list, so "trample" is a fact not a guess
    oracle text    0.35   how much of what the old card said the new one says

then a popularity nudge of up to 0.10 from EDHREC rank, as a tiebreak only. It
is last and it is small on purpose: popular is not the same as correct, and a
scorer that leads with it just recommends staples.
"""

from __future__ import annotations

import math
import re
from dataclasses import dataclass

from mtglab.cards import db
from mtglab.cards.db import CardRecord, Connection
from mtglab.decks.model import Deck

# Card types, most specific first: a "Legendary Artifact Creature" is a
# creature, and matching it against an artifact would be the wrong call.
PRIMARY_TYPES = (
    "Creature", "Planeswalker", "Land", "Instant", "Sorcery",
    "Artifact", "Enchantment", "Battle",
)

PERMANENT_TYPES = frozenset(
    {"Creature", "Planeswalker", "Land", "Artifact", "Enchantment", "Battle"})

# English function words, plus the Magic templating that appears on almost
# every card and therefore distinguishes nothing. Kept as prose rather than a
# list literal (SIM905) because this is a word list people will edit, and a
# hundred quoted, comma-separated strings is markedly harder to scan and to
# diff than five lines of words.
STOPWORDS = frozenset("""
a an and as at be been but by can cant do does each end for from has have if in
into is it its may more must not of on onto or other others out over own put
same than that the their them then there these they this those to under until
up upon was were when whenever where which while with without you your
also another any are back both get gets much no only rather very
""".split())  # noqa: SIM905

_WORD = re.compile(r"[a-z]+")


@dataclass(frozen=True)
class Candidate:
    """One suggestion, with the arithmetic that produced it."""

    record: CardRecord
    score: float
    reasons: tuple[str, ...]

    @property
    def name(self) -> str:
        return self.record.name


def primary_type(type_line: str) -> str:
    """The card's main type, read from the front face.

    Front face only, for the same reason `CardRecord.is_land` reads it: a
    modal DFC's combined type line names both faces, and matching a creature
    against "Creature — Elf // Land" on the word "Land" is how a spell ends up
    suggested as a land.
    """
    front = (type_line or "").split(" // ")[0]
    for candidate in PRIMARY_TYPES:
        if candidate in front:
            return candidate
    return ""


def _tokens(*texts: str) -> set[str]:
    """Significant words, lowercased, stopwords and short words removed."""
    out: set[str] = set()
    for text in texts:
        for word in _WORD.findall((text or "").lower()):
            if len(word) > 2 and word not in STOPWORDS:
                out.add(word)
    return out


def _type_score(target: CardRecord, candidate: CardRecord) -> float:
    left, right = primary_type(target.type_line), primary_type(candidate.type_line)
    if not left or not right:
        return 0.0
    if left == right:
        return 1.0
    # A permanent for a permanent is a partial match: it still occupies a slot
    # on the battlefield, which is most of what a deck's shape cares about.
    if left in PERMANENT_TYPES and right in PERMANENT_TYPES:
        return 0.4
    return 0.0


def _curve_score(target: CardRecord, candidate: CardRecord) -> float:
    return max(0.0, 1.0 - abs(target.cmc - candidate.cmc) / 3.0)


def _keyword_score(target: CardRecord, candidate: CardRecord) -> float:
    left = {k.lower() for k in target.keywords}
    right = {k.lower() for k in candidate.keywords}
    if not left:
        # Nothing to match against. Scoring 0 rather than renormalising is fine
        # because it is 0 for every candidate, so the ranking is unaffected.
        return 0.0
    return len(left & right) / len(left | right)


def _text_score(target_tokens: set[str], candidate: CardRecord) -> float:
    """How much of what the old card said the new one also says.

    Asymmetric on purpose -- the denominator is the *target's* vocabulary, so a
    long card that happens to contain the target's words scores well, and a
    short card is not rewarded merely for being short.
    """
    if not target_tokens:
        return 0.0
    shared = target_tokens & _tokens(candidate.oracle_text)
    return len(shared) / len(target_tokens)


def _popularity(rec: CardRecord) -> float:
    """EDHREC rank, log-scaled into 0..1. Unranked scores 0.

    Note a banned card is itself unranked, which is a small reminder that this
    number measures what people play, not what is good.
    """
    if not rec.edhrec_rank or rec.edhrec_rank < 1:
        return 0.0
    return max(0.0, 1.0 - math.log10(rec.edhrec_rank) / 5.0)


def score(target: CardRecord, candidate: CardRecord, *, why: str = "") -> Candidate:
    """Score one candidate against the card being replaced.

    `why` is the deck's own rationale for the slot. It is folded into the text
    comparison because it says what the card was *for*, which the oracle text
    often does not -- "ramp and threat in one card" is not a phrase that
    appears on Primeval Titan.
    """
    target_tokens = _tokens(target.oracle_text, why)

    parts = (
        (0.30, _type_score(target, candidate)),
        (0.20, _curve_score(target, candidate)),
        (0.15, _keyword_score(target, candidate)),
        (0.35, _text_score(target_tokens, candidate)),
    )
    similarity = sum(weight * value for weight, value in parts)
    total = similarity + 0.10 * _popularity(candidate)

    reasons: list[str] = []
    if _type_score(target, candidate) == 1.0:
        reasons.append(f"same card type ({primary_type(candidate.type_line)})")
    if candidate.cmc == target.cmc:
        reasons.append(f"same mana value ({int(target.cmc)})")
    elif abs(candidate.cmc - target.cmc) <= 1:
        reasons.append(f"mana value {int(candidate.cmc)} vs {int(target.cmc)}")
    shared_keywords = ({k.lower() for k in target.keywords}
                       & {k.lower() for k in candidate.keywords})
    if shared_keywords:
        reasons.append("shares " + ", ".join(sorted(shared_keywords)))
    shared_words = sorted(target_tokens & _tokens(candidate.oracle_text))
    if shared_words:
        reasons.append("text: " + ", ".join(shared_words[:6]))
    if candidate.edhrec_rank:
        reasons.append(f"EDHREC rank {candidate.edhrec_rank:,}")

    return Candidate(record=candidate, score=round(total, 4),
                     reasons=tuple(reasons))


def rank(target: CardRecord, candidates: list[CardRecord], *, why: str = "",
         limit: int = 5, exclude: frozenset[str] = frozenset()) -> list[Candidate]:
    """Score every candidate and return the best `limit`.

    Ties break by name, so the output is stable run to run -- a shortlist that
    reshuffles itself between two identical runs is one nobody trusts.
    """
    blocked = {n.lower() for n in exclude} | {target.name.lower()}
    scored = [
        score(target, rec, why=why)
        for rec in candidates
        if rec.name.lower() not in blocked
    ]
    scored.sort(key=lambda c: (-c.score, c.name))
    return scored[:limit]


def candidate_pool(con: Connection, target: CardRecord,
                   identity: frozenset[str], *,
                   pool_size: int = 400) -> list[CardRecord]:
    """Everything legal that could plausibly fill the slot.

    A prefilter, not a ranking: it narrows ~35k cards to a few hundred that are
    legal, inside the commander's identity, the right sort of card and roughly
    the right cost, and leaves the judgement to `rank`.
    """
    colors = sorted(c for c in identity if c in "WUBRG")
    listed = ", ".join(f"'{c}'" for c in colors) or "''"

    where = [
        "json_extract_string(legalities, 'commander') = 'legal'",
        f"len(list_filter(color_identity, x -> x NOT IN ({listed}))) = 0",
        "cmc BETWEEN ? AND ?",
    ]
    params: list[object] = [max(0.0, target.cmc - 2), target.cmc + 2]

    kind = primary_type(target.type_line)
    if kind:
        where.append("type_line LIKE ?")
        params.append(f"%{kind}%")

    return db.search(con, " AND ".join(where), params, limit=pool_size,
                     order_by="edhrec_rank NULLS LAST")


def replacements_for(deck: Deck, cards: dict[str, CardRecord], con: Connection,
                     name: str, *,
                     limit: int = 5) -> list[Candidate]:
    """Suggestions for one card in one deck.

    Returns an empty list rather than raising when the corpus does not know the
    card -- the gate has already said so, and failing twice for one problem
    helps nobody.
    """
    target = cards.get(name)
    if target is None:
        return []

    identity: frozenset[str] = frozenset()
    for commander in deck.commander:
        rec = cards.get(commander)
        if rec is not None:
            identity |= rec.color_identity

    why = next((c.why for c in deck.cards if c.name == name), "")
    # Never suggest something already in the list, or on the swap board.
    already = frozenset(
        [c.name for c in deck.cards] + [c.name for c in deck.swap_board] + deck.commander,
    )
    pool = candidate_pool(con, target, identity)
    return rank(target, pool, why=why, limit=limit, exclude=already)

File: go/internal/suggest/suggest.go
// Package suggest is `decks/suggest.py`: ranked replacement candidates for a
// card that has to leave a deck -- a measurement, not a recommendation. Of
// every card that is legal here, which most resemble the one being removed?
// A weighted sum (type 0.30, mana value 0.20, keywords 0.15, oracle text
// 0.35) and a popularity nudge of up to 0.10 from EDHREC rank, last and
// small on purpose: popular is not the same as correct. Every fact comes
// from the pool; nothing here decides anything.
package suggest

import (
	"context"
	"fmt"
	"math"
	"regexp"
	"sort"
	"strings"

	"github.com/aasquier/sylvan-library/go/internal/deck"
	"github.com/aasquier/sylvan-library/go/internal/pool"
)

// PrimaryTypes are the card types, most specific first: a "Legendary
// Artifact Creature" is a creature.
var PrimaryTypes = []string{"Creature", "Planeswalker", "Land", "Instant", "Sorcery", "Artifact", "Enchantment", "Battle"}

var permanentTypes = map[string]bool{"Creature": true, "Planeswalker": true, "Land": true,
	"Artifact": true, "Enchantment": true, "Battle": true}

// Stopwords are English function words plus the Magic templating that
// appears on almost every card and therefore distinguishes nothing.
var Stopwords = map[string]bool{}

func init() {
	for _, w := range strings.Fields(`
a an and as at be been but by can cant do does each end for from has have if in
into is it its may more must not of on onto or other others out over own put
same than that the their them then there these they this those to under until
up upon was were when whenever where which while with without you your
also another any are back both get gets much no only rather very
`) {
		Stopwords[w] = true
	}
}

var wordRe = regexp.MustCompile(`[a-z]+`)

// Candidate is one suggestion, with the arithmetic that produced it.
type Candidate struct {
	Record  *pool.CardRecord
	Score   float64
	Reasons []string
}

// Name is the candidate's.
func (c Candidate) Name() string { return c.Record.Name }

// PrimaryType is `suggest.primary_type`: the card's main type, from the
// front face.
func PrimaryType(typeLine string) string {
	front, _, _ := strings.Cut(typeLine, " // ")
	for _, t := range PrimaryTypes {
		if strings.Contains(front, t) {
			return t
		}
	}
	return ""
}

// Tokens is `_tokens`: significant words, lowercased, stopwords and short
// words removed.
func Tokens(texts ...string) map[string]bool {
	out := map[string]bool{}
	for _, text := range texts {
		for _, w := range wordRe.FindAllString(strings.ToLower(text), -1) {
			if len(w) > 2 && !Stopwords[w] {
				out[w] = true
			}
		}
	}
	return out
}

func typeScore(target, candidate *pool.CardRecord) float64 {
	left, right := PrimaryType(target.TypeLine), PrimaryType(candidate.TypeLine)
	if left == "" || right == "" {
		return 0
	}
	if left == right {
		return 1
	}
	if permanentTypes[left] && permanentTypes[right] {
		return 0.4
	}
	return 0
}

func curveScore(target, candidate *pool.CardRecord) float64 {
	return math.Max(0, 1-math.Abs(target.CMC-candidate.CMC)/3)
}

func lowerSet(list []string) map[string]bool {
	out := map[string]bool{}
	for _, s := range list {
		out[strings.ToLower(s)] = true
	}
	return out
}

func keywordScore(target, candidate *pool.CardRecord) float64 {
	left, right := lowerSet(target.Keywords), lowerSet(candidate.Keywords)
	if len(left) == 0 {
		return 0
	}
	both, either := 0, len(left)
	for k := range right {
		if left[k] {
			both++
		} else {
			either++
		}
	}
	return float64(both) / float64(either)
}

func textScore(targetTokens map[string]bool, candidate *pool.CardRecord) float64 {
	if len(targetTokens) == 0 {
		return 0
	}
	cand := Tokens(candidate.OracleText)
	shared := 0
	for w := range targetTokens {
		if cand[w] {
			shared++
		}
	}
	return float64(shared) / float64(len(targetTokens))
}

// popularity is `_popularity`: EDHREC rank, log-scaled into 0..1.
func popularity(rec *pool.CardRecord) float64 {
	if rec.EdhrecRank == nil || *rec.EdhrecRank < 1 {
		return 0
	}
	return math.Max(0, 1-math.Log10(float64(*rec.EdhrecRank))/5)
}

// round4 is Python's `round(x, 4)`.
func round4(x float64) float64 { return math.RoundToEven(x*10000) / 10000 }

// thousands is Python's `f"{n:,}"`.
func thousands(n int) string {
	s := fmt.Sprint(n)
	if len(s) <= 3 {
		return s
	}
	var b strings.Builder
	pre := len(s) % 3
	if pre > 0 {
		b.WriteString(s[:pre])
	}
	for i := pre; i < len(s); i += 3 {
		if b.Len() > 0 {
			b.WriteByte(',')
		}
		b.WriteString(s[i : i+3])
	}
	return b.String()
}

// Score is `suggest.score`: one candidate against the card being replaced.
// `why` is the deck's own rationale for the slot, folded into the text
// comparison because it says what the card was *for*.
func Score(target, candidate *pool.CardRecord, why string) Candidate {
	targetTokens := Tokens(target.OracleText, why)
	similarity := 0.30*typeScore(target, candidate) + 0.20*curveScore(target, candidate) +
		0.15*keywordScore(target, candidate) + 0.35*textScore(targetTokens, candidate)
	total := similarity + 0.10*popularity(candidate)

	reasons := []string{}
	if typeScore(target, candidate) == 1 {
		reasons = append(reasons, fmt.Sprintf("same card type (%s)", PrimaryType(candidate.TypeLine)))
	}
	if candidate.CMC == target.CMC {
		reasons = append(reasons, fmt.Sprintf("same mana value (%d)", int(target.CMC)))
	} else if math.Abs(candidate.CMC-target.CMC) <= 1 {
		reasons = append(reasons, fmt.Sprintf("mana value %d vs %d", int(candidate.CMC), int(target.CMC)))
	}
	left, right := lowerSet(target.Keywords), lowerSet(candidate.Keywords)
	shared := []string{}
	for k := range left {
		if right[k] {
			shared = append(shared, k)
		}
	}
	if len(shared) > 0 {
		sort.Strings(shared)
		reasons = append(reasons, "shares "+strings.Join(shared, ", "))
	}
	cand := Tokens(candidate.OracleText)
	words := []string{}
	for w := range targetTokens {
		if cand[w] {
			words = append(words, w)
		}
	}
	if len(words) > 0 {
		sort.Strings(words)
		if len(words) > 6 {
			words = words[:6]
		}
		reasons = append(reasons, "text: "+strings.Join(words, ", "))
	}
	if candidate.EdhrecRank != nil && *candidate.EdhrecRank != 0 {
		reasons = append(reasons, "EDHREC rank "+thousands(*candidate.EdhrecRank))
	}
	return Candidate{Record: candidate, Score: round4(total), Reasons: reasons}
}

// Rank is `suggest.rank`: score every candidate and return the best `limit`,
// ties broken by name so the output is stable run to run.
func Rank(target *pool.CardRecord, candidates []*pool.CardRecord, why string, limit int, exclude map[string]bool) []Candidate {
	blocked := map[string]bool{strings.ToLower(target.Name): true}
	for n := range exclude {
		blocked[strings.ToLower(n)] = true
	}
	scored := []Candidate{}
	for _, rec := range candidates {
		if blocked[strings.ToLower(rec.Name)] {
			continue
		}
		scored = append(scored, Score(target, rec, why))
	}
	sort.SliceStable(scored, func(i, j int) bool {
		if scored[i].Score != scored[j].Score {
			return scored[i].Score > scored[j].Score
		}
		return scored[i].Name() < scored[j].Name()
	})
	if len(scored) > limit {
		scored = scored[:limit]
	}
	return scored
}

// CandidatePool is `suggest.candidate_pool`: everything legal that could
// plausibly fill the slot -- a prefilter to a few hundred, inside the
// identity, roughly the right cost and sort of card.
func CandidatePool(ctx context.Context, c *pool.Conn, target *pool.CardRecord, identity map[string]bool, poolSize int) ([]*pool.CardRecord, error) {
	colors := []string{}
	for _, col := range []string{"W", "U", "B", "R", "G"} {
		if identity[col] {
			colors = append(colors, col)
		}
	}
	listed := "''"
	if len(colors) > 0 {
		quoted := make([]string, len(colors))
		for i, col := range colors {
			quoted[i] = "'" + col + "'"
		}
		listed = strings.Join(quoted, ", ")
	}
	where := []string{
		"json_extract_string(legalities, 'commander') = 'legal'",
		fmt.Sprintf("len(list_filter(color_identity, x -> x NOT IN (%s))) = 0", listed),
		"cmc BETWEEN ? AND ?",
	}
	params := []any{math.Max(0, target.CMC-2), target.CMC + 2}
	if kind := PrimaryType(target.TypeLine); kind != "" {
		where = append(where, "type_line LIKE ?")
		params = append(params, "%"+kind+"%")
	}
	return c.Search(ctx, strings.Join(where, " AND "), params, poolSize, "edhrec_rank NULLS LAST", 0)
}

// ReplacementsFor is `suggest.replacements_for`: suggestions for one card in
// one deck; an empty list when the pool does not know the card.
func ReplacementsFor(ctx context.Context, c *pool.Conn, d *deck.Deck, cards map[string]*pool.CardRecord, name string, limit int) ([]Candidate, error) {
	target := cards[name]
	if target == nil {
		return []Candidate{}, nil
	}
	identity := map[string]bool{}
	for _, commander := range d.Commander {
		if rec := cards[commander]; rec != nil {
			for _, col := range rec.ColorIdentity {
				identity[col] = true
			}
		}
	}
	why := ""
	for _, entry := range d.Cards {
		if entry.Name == name {
			why = entry.Why
			break
		}
	}
	already := map[string]bool{}
	for _, entry := range d.Cards {
		already[entry.Name] = true
	}
	for _, entry := range d.SwapBoard {
		already[entry.Name] = true
	}
	for _, commander := range d.Commander {
		already[commander] = true
	}
	candidates, err := CandidatePool(ctx, c, target, identity, 400)
	if err != nil {
		return nil, err
	}
	return Rank(target, candidates, why, limit, already), nil
}
